File: algorithms/0543.diameter-of-binary-tree/solution.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer}
def diameter_of_binary_tree(root)
  @longest = 0
  depth(root) unless root.nil?
  @longest
end

def depth(node)
  return 0 if node.nil?

  left = depth(node.left)
  right = depth(node.right)
  @longest = [@longest, left + right].max

  [left, right].max + 1
end

File: algorithms/0543.diameter-of-binary-tree/solution.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;

type Node = Option<Rc<RefCell<TreeNode>>>;

impl Solution {
    pub fn diameter_of_binary_tree(root: Node) -> i32 {
        let mut longest = 0;
        Self::depth(root, &mut longest);
        longest
    }

    fn depth(node: Node, longest: &mut i32) -> i32 {
        if let Some(node) = node {
            let left = Self::depth(node.borrow().left.clone(), longest);
            let right = Self::depth(node.borrow().right.clone(), longest);
            *longest = std::cmp::max(*longest, left + right);
            1 + left.max(right)
        } else {
            0
        }
    }
}
